### PR TITLE
FARM-Beta validator

### PR DESCRIPTION
**Pull request by wanghy-anl**
--------
Pull Request Description
--------
##### What issue does this change request address?
FARM based validator is implemented. 
- FARM-Beta is the new version ready for production phase; Also it will be the ground brick for FARM-Gamma that will be developed by April 2022.
- FARM-Alpha is the prototype where everything is hard-coded. FARM-Alpha can be deleted to save reviewer's time.

##### What are the significant changes in functionality due to this change request?
The FARM-Beta validator is implemented in this pull request. FARM-Beta recognizes the user-provided IES list of components and system configurations from the HERON input file, and perform the setpoint validation during the power dispatch optimization so that the system implicit operational constraints are not violated during transients and steady-state operations.

----------------
For Change Control Board: Change Request Review
----------------
The following review must be completed by an authorized member of the Change Control Board.
- [ ] 1. Review all computer code.
- [ ] 2. If any changes occur to the input syntax, there must be an accompanying change to the user manual and xsd schema. If the input syntax change deprecates existing input files, a conversion script needs to be added (see Conversion Scripts).
- [ ] 3. Make sure the Python code and commenting standards are respected (camelBack, etc.) - See on the [wiki](https://github.com/idaholab/HERON/wiki/Code-Standards) for details.
- [ ] 4. Automated Tests should pass.
- [ ] 5. If significant functionality is added, there must be tests added to check this. Tests should cover all possible options.  Multiple short tests are preferred over one large tes.
- [ ] 6. If the change modifies or adds a requirement or a requirement based test case, the Change Control Board's Chair or designee also needs to approve the change.  The requirements and the requirements test shall be in sync.
- [ ] 7. The merge request must reference an issue.  If the issue is closed, the issue close checklist shall be done.
- [ ] 8. If an analytic test is changed/added, the the analytic documentation must be updated/added.
- [ ] 9. If any test used as a basis for documentation examples have been changed, the associated documentation must be reviewed and assured the text matches the example.

### CODE REVIEW

File: src/validators/FARMValidators.py [hunk 1]
```diff
@@ -0,0 +1,915 @@
...
+    self.name = 'BaseValidator'
+    self._tolerance = 1.003e-6
+
+  def read_input(self, inputs):
+    """
+      Loads settings based on provided inputs
+      @ In, inputs, InputData.InputSpecs, input specifications
+      @ Out, None
+    """
+    print("Haoyu TEST: read_input\n")
```
Comment (dylanjm) [inline src/validators/FARMValidators.py]: I would remove all print statements or use our MessageHandler to raise debug messages. There are a few print statements throughout this file.

File: src/validators/_utils.py [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+import importlib
+import xml.etree.ElementTree as ET
+
+def get_heron_loc():
+  """
+    Return HERON location
+    @ In, None
+    @ Out, loc, string, absolute location of HERON
+  """
+  return os.path.abspath(os.path.join(__file__, '..', '..'))
```
Comment (PaulTalbot-INL) [inline src/validators/_utils.py]: since this is within HERON, perhaps using the already-existing HERON utils for this would work instead of copying this file?

File: tests/integration_tests/mechanics/validator_farm_alpha/cash.xml [hunk 3]
```diff
@@ -0,0 +1,52 @@
+<Economics verbosity="50">
```
Comment (PaulTalbot-INL) [inline tests/integration_tests/mechanics/validator_farm_alpha/cash.xml]: This is a generated file and shouldn't be included in the PR.

File: tests/integration_tests/mechanics/validator_farm_beta/cash.xml [hunk 4]
```diff
@@ -0,0 +1,52 @@
+<Economics verbosity="50">
```
Comment (PaulTalbot-INL) [inline tests/integration_tests/mechanics/validator_farm_beta/cash.xml]: generated file, should not be in PR

File: tests/integration_tests/mechanics/validator_farm_beta/write_inner.py [hunk 5]
```diff
@@ -0,0 +1,27 @@
+import xml.etree.ElementTree as ET
```
Comment (PaulTalbot-INL) [inline tests/integration_tests/mechanics/validator_farm_beta/write_inner.py]: generated file, should not be in PR

File: tests/integration_tests/mechanics/validator_farm_alpha/write_inner.py [hunk 6]
```diff
@@ -0,0 +1,27 @@
+import xml.etree.ElementTree as ET
```
Comment (PaulTalbot-INL) [inline tests/integration_tests/mechanics/validator_farm_alpha/write_inner.py]: generated file

**Comment by wanghy-anl** [comment]
@dylanjm @PaulTalbot-INL Hey Dylan and Paul, would you place the review work on hold for one day or two? I will let you know when I addressed your comments in one day or two. We are overwhelmed by the January Milestone this week. Thanks, Haoyu

**Comment by wanghy-anl** [comment]
@PaulTalbot-INL @dylanjm Hey Dylan and Paul, I have deleted the generated files and basically all the "print" commands. Also, the "_utils.py" import has been fixed. 
However I saw both the Precheck and CentOS 8 check has failed. Could you share with me the error messages? When I click the "Details" link it says "You are not allowed to see results. Please sign in to an appropriate user." 
Thanks! Haoyu

File: tests/integration_tests/mechanics/validator_farm_alpha/write_inner.py (hunk 6 above)
Comment (wanghy-anl) [reply tests/integration_tests/mechanics/validator_farm_alpha/write_inner.py]: deleted

File: tests/integration_tests/mechanics/validator_farm_beta/write_inner.py (hunk 5 above)
Comment (wanghy-anl) [reply tests/integration_tests/mechanics/validator_farm_beta/write_inner.py]: deleted

File: tests/integration_tests/mechanics/validator_farm_beta/cash.xml (hunk 4 above)
Comment (wanghy-anl) [reply tests/integration_tests/mechanics/validator_farm_beta/cash.xml]: deleted

File: tests/integration_tests/mechanics/validator_farm_alpha/cash.xml (hunk 3 above)
Comment (wanghy-anl) [reply tests/integration_tests/mechanics/validator_farm_alpha/cash.xml]: deleted

File: src/validators/_utils.py (hunk 2 above)
Comment (wanghy-anl) [reply src/validators/_utils.py]: fixed

File: src/validators/FARMValidators.py (hunk 1 above)
Comment (wanghy-anl) [reply src/validators/FARMValidators.py]: I deleted basically all the print commands

**Comment by PaulTalbot-INL** [comment]
> Hey Dylan and Paul, I have deleted the generated files and basically all the "print" commands. Also, the "_utils.py" import has been fixed.
> However I saw both the Precheck and CentOS 8 check has failed. Could you share with me the error messages? When I click the "Details" link it says "You are not allowed to see results. Please sign in to an appropriate user."
> Thanks! Haoyu

I assume you tried signing in using GitHub? Does it still not show you results?

CentOS 8 failed because precheck failed, and precheck failed with:
```
##########################################################################
ERROR: The following files contain trailing whitespace after applying your patch:
	src/validators/FARMValidators.py
	tests/integration_tests/mechanics/validator_farm_beta/transfers.py

Run the "delete_trailing_whitespace.sh" script in your $MOOSE_DIR/scripts directory.
##########################################################################


ERROR: exiting with code 1


//: Singularity execution completed with return code 1
```

**Comment by wanghy-anl** [comment]
@PaulTalbot-INL Thank you Paul for the information. You are correct, I have signed in on GitHub, but it doesn't show me the results. 
I ran the "delete_trailing_whitespace.sh" on these two files and now the precheck is passed.
Thanks,
Haoyu

**Comment by PaulTalbot-INL** [comment]
Superseded by #150.
